File: example/fs/Samples/STMTest.fs
namespace Samples

open System
open AUTD3Sharp
open AUTD3Sharp.Gain
open AUTD3Sharp.Modulation
open AUTD3Sharp.Utils
open System.Threading.Tasks
open type AUTD3Sharp.Units

module STMTest =
    let GainSTMTest<'T> (autd : Controller<'T>) = 
        (Silencer.Disable()) |> autd.Send;

        (new Static()) |> autd.Send;
        
        let center = autd.Geometry.Center + Vector3d(0, 0, 150);
        let stm = 
            [0..199]
            |> List.map (fun i -> (2.0 * Math.PI * (float)i / 200.0))
            |> List.map (fun theta -> (center + 30.0 * Vector3d(cos(theta), sin(theta), 0.0)))
            |> List.map (fun p -> (new Focus(p)))
            |> List.fold (fun (acc: GainSTM) v -> acc.AddGain v) (GainSTM.FromFreq(1.0 * Hz))

        autd.Send stm;

    let FocusSTMTest<'T> (autd : Controller<'T>) = 
        (Silencer.Disable()) |> autd.Send;

        (new Static()) |> autd.Send;
        
        let center = autd.Geometry.Center + Vector3d(0, 0, 150);
        let stm = 
            [0..199]
            |> List.map (fun i -> (2.0 * Math.PI * (float)i / 200.0))
            |> List.map (fun theta -> (center + 30.0 * Vector3d(cos(theta), sin(theta), 0.0)))
            |> List.fold (fun (acc: FocusSTM) v -> acc.AddFocus v) (FocusSTM.FromFreq(1.0 * Hz))

        autd.Send stm;

File: example/Samples/STMTest.cs
using AUTD3Sharp;
using AUTD3Sharp.Gain;
using AUTD3Sharp.Modulation;
using AUTD3Sharp.Utils;
using static AUTD3Sharp.Units;

namespace Samples;

internal static class GainSTMTest
{
    public static void Test(Controller autd)
    {
        var config = Silencer.Disable();
        autd.Send(config);

        var m = new Static();

        var center = autd.Center() + new Vector3(0, 0, 150);
        const int pointNum = 50;
        const float radius = 30.0f;
        var stm = new GainSTM(gains: Enumerable.Range(0, pointNum).Select(i =>
        {
            var theta = 2.0f * MathF.PI * i / pointNum;
            return new Focus(pos: center + radius * new Vector3(MathF.Cos(theta), MathF.Sin(theta), 0), option: new FocusOption());
        }), config: 1.0f * Hz, option: new GainSTMOption());

        autd.Send((m, stm));
    }
}

internal static class FociSTMTest
{
    public static void Test(Controller autd)
    {
        var config = Silencer.Disable();
        autd.Send(config);

        var mod = new Static();
        autd.Send(mod);

        var center = autd.Center() + new Vector3(0, 0, 150);
        const int pointNum = 200;
        const float radius = 30.0f;
        var stm = new FociSTM(foci: Enumerable.Range(0, pointNum).Select(i =>
        {
            var theta = 2.0f * MathF.PI * i / pointNum;
            return center + radius * new Vector3(MathF.Cos(theta), MathF.Sin(theta), 0);
        }), config: 1.0f * Hz);

        autd.Send(stm);
    }
}
